environment/template.go

```diff
@@ -0,0 +1,281 @@
+	ValKindList                   = "List"
+	ValKindDeployment             = "Deployment"
+	ValKindDeploymentConfig       = "DeploymentConfig"
+	ValKindResourceQuota          = "ResourceQuota"
+
+	varUserName              = "USER_NAME"
+	varProjectUser           = "PROJECT_USER"
+	varProjectRequestingUser = "PROJECT_REQUESTING_USER"
+	varProjectAdminUser      = "PROJECT_ADMIN_USER"
+	varKeycloakURL           = "KEYCLOAK_URL"
+	varCommit                = "COMMIT"
+	varKeycloakOsoEndpoint   = "KEYCLOAK_OSO_ENDPOINT"
+	varKeycloakGHEndpoint    = "KEYCLOAK_GITHUB_ENDPOINT"
+)
+
+var sortOrder = map[string]int{
+	"Namespace":              1,
+	"ProjectRequest":         1,
+	"RoleBindingRestriction": 2,
+	"LimitRange":             3,
+	"ResourceQuota":          4,
+	"Secret":                 5,
+	"ServiceAccount":         6,
+	"Service":                7,
+	"RoleBinding":            8,
+	"PersistentVolumeClaim":  9,
+	"ConfigMap":              10,
+	"DeploymentConfig":       11,
+	"Route":                  12,
+	"Job":                    13,
+}
+
+type Objects []map[interface{}]interface{}
+type Object map[interface{}]interface{}
+
+type Template struct {
+	Filename      string
+	DefaultParams map[string]string
+	Content       string
+	Version       string
+}
+
+var (
+	stageParams = map[string]string{"DEPLOY_TYPE": "stage"}
+	runParams   = map[string]string{"DEPLOY_TYPE": "run"}
+	noParams    map[string]string
+)
+
+func newTemplate(filename string, defaultParams map[string]string) *Template {
+	return &Template{
+		Filename:      filename,
+		DefaultParams: defaultParams,
+	}
+}
+
+func (t *Template) Process(vars map[string]string) (Objects, error) {
+	var objects Objects
+	templateVars := merge(vars, t.DefaultParams)
+	pt, err := t.ReplaceVars(templateVars)
+	if err != nil {
+		return objects, err
+	}
+	t.Version = vars[varCommit]
+	return ParseObjects(pt)
+}
+
+// Process takes a K8/Openshift Template as input and resolves the variable expresions
+func (t *Template) ReplaceVars(variables map[string]string) (string, error) {
+	reg := regexp.MustCompile(`\${([A-Z_0-9]+)}`)
+	return string(reg.ReplaceAllFunc([]byte(t.Content), func(found []byte) []byte {
+		variableName := toVariableName(string(found))
+		if variable, ok := variables[variableName]; ok {
+			return []byte(variable)
+		}
+		return found
+	})), nil
+}
+
+func CollectVars(user, masterUser, commit string, config *configuration.Data) map[string]string {
+	userName := RetrieveUserName(user)
+
+	vars := map[string]string{
+		varUserName:              userName,
+		varProjectUser:           user,
+		varProjectRequestingUser: user,
+		varProjectAdminUser:      masterUser,
+		varCommit:                commit,
+	}
+
+	return merge(vars, getVariables(config))
+}
+
+// RetrieveUserName returns a safe namespace basename based on a username
+func RetrieveUserName(username string) string {
+	return regexp.MustCompile("[^a-z0-9]").ReplaceAllString(strings.Split(username, "@")[0], "-")
```

Comment:
`regexp.MustCompile("[^a-z0-9]")` need to run only once so can we move as var outside func.